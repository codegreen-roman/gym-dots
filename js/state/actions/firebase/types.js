export const GOT_DEFAULTS = 'GOT_DEFAULTS'
export const GOT_ERROR_LOADING_DEFAULTS = 'GOT_ERROR_GETTING_DEFAULTS'
export const START_AUTH = 'START_AUTH'
export const COMPLETE_AUTH = 'COMPLETE_AUTH'
export const AUTH_SUCCESS = 'AUTH_SUCCESS'
export const AUTH_ERROR = 'AUTH_ERROR'
export const EXERCISES_FETCHING_SUCCESS = ' EXERCISES_FETCHING_SUCCESS'
export const EXERCISES_FETCHING_ERROR = ' EXERCISES_FETCHING_ERROR'
