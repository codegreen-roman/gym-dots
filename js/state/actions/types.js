export const TEST_ACTION = 'TEST_ACTION'
export const GET_LOCATION = 'GET_LOCATION'
export const GET_LOCATION_START = 'GET_LOCATION_START'
export const PING = 'PING'
export const LOAD_SESSIONS_SUCCESS = 'LOAD_SESSIONS_SUCCESS'

