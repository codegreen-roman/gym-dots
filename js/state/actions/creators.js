import { PING, GET_LOCATION, GET_LOCATION_START, WORKOUT_STATUS } from './types'
import { compose } from 'ramda'

export const startWorkout = () => ({
    type: WORKOUT_STATUS,
    payload: {
        status: 'starting'
    }
})

export const setStartedWorkout = () => ({
    type: WORKOUT_STATUS,
    payload: {
        status: 'started'
    }
})


export const ping = payload => ({
    type: PING,
    payload
})

const loadLocationSuccess = location => ({
    type: GET_LOCATION,
    payload: {
        location,
        status: 'ready'
    }
})

const loadLocationError = () => ({
    type: GET_LOCATION,
    payload: {
        status: 'failed'
    }
})

export const loadLocationStart = () => ({
    type: GET_LOCATION_START,
    payload: {
        status: 'started'
    }
})

export const startWorkoutWithCountdown = () => {
    return function (dispatch) {

        dispatch(startWorkout())

        const promise = new Promise((resolve) => {
            setTimeout(() => {
                resolve('started')
            }, 3000)
        })

        return promise
            .then(compose(dispatch, setStartedWorkout))
    }
}

export const loadLocationForUsername = () => {
    return function (dispatch) {

        dispatch(loadLocationStart())

        const promise = new Promise((resolve) => {
            setTimeout(() => {
                resolve('Tallinn, Estonia')
            }, 1000)
        })

        return promise
            .then(compose(dispatch, loadLocationSuccess))
    }
}
