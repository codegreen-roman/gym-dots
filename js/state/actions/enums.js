export const actions = {
    'TEST_ACTION': 'TEST_ACTION'
}
