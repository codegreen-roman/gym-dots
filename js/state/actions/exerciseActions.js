import { SET_FAILED, SET_SUCCESSFULLY_COMPLETED } from './types'

export const setFailed = () => {
    return {
        type: SET_FAILED,
        payload: {}
    }
}

export const setDone = () => {
    return {
        type: SET_SUCCESSFULLY_COMPLETED,
        payload: {}
    }
}
