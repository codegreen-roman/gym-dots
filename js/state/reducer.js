import { combineReducers } from 'redux'
import { i18nReducer } from 'react-redux-i18n'
import { exerciseSession, githubLocation, sessions, workoutStatus, defaults, auth } from './reducers'

export const reducer = combineReducers({
    exerciseSession,
    githubLocation,
    sessions,
    workoutStatus,
    defaults,
    auth,
    i18n: i18nReducer
})
