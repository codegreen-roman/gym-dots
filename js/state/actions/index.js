import { ping, startWorkoutWithCountdown, setIntermediateWorkout } from './creators'
import { loadSessions } from './session-actions'
import { exercisesOrderChange, moveExerciseToCompleted } from './exercisesActions'
import {
    loadAppDefaults,
    subscribeToAppDefaultsChanges,
    subscribeToAuthStateChanged,
    saveExercisesResults
} from './firebase/databaseActions'

export {
    ping,
    loadSessions,
    startWorkoutWithCountdown,
    loadAppDefaults,
    subscribeToAppDefaultsChanges,
    exercisesOrderChange,
    setIntermediateWorkout,
    moveExerciseToCompleted,
    subscribeToAuthStateChanged,
    saveExercisesResults
}
