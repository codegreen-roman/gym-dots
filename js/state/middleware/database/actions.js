import { compose, ifElse, isNil } from 'ramda'
import { types } from '../../'
import { getVal } from './actions.helper'
import { push } from 'react-router-redux'

const {
    AUTH_SUCCESS,
    AUTH_VOID,
    AUTH_ERROR,
    EXERCISES_FETCHING_ERROR,
    EXERCISES_FETCHING_SUCCESS
} = types

import { loginWith, loginAnonymously, logout, auth, loadNextSessionForUser } from './database'

const gotErrorWhileAuth = ({ code, message, email, credential }) => ({
    type: AUTH_ERROR,
    payload: { code, message, email, credential }
})

const authSuccess = ({ user }) => ({
    type: AUTH_SUCCESS,
    payload: { user }
})

const authVoid = () => ({
    type: AUTH_VOID,
    payload: {}
})

const exercisesFetchingSuccess = nextSession => ({
    type: EXERCISES_FETCHING_SUCCESS,
    payload: {
        nextSession
    }
})

const exercisesFetchingError = error => ({
    type: EXERCISES_FETCHING_ERROR,
    payload: {
        error
    }
})

const authActionMaker = (method) => (provider) => (dispatch) => method(provider).catch(compose(dispatch, gotErrorWhileAuth))

export const authWith = authActionMaker(loginWith)
export const authAnonymously = authActionMaker(loginAnonymously)
export const authVoidAction = authActionMaker(logout)

export const subscribeToAuthStateChanged = dispatch => {

    const passUser = compose(dispatch, authSuccess)
    // const navigateHome = compose(dispatch, push('/'))
    const loggedOut = compose(dispatch, authVoid)

    return new Promise(resolve => {
        auth.onAuthStateChanged(function (user) {
            if (user) {
                passUser({ user })
                loadNextSession(user.uid)(dispatch)
                resolve({ user })
            } else {
                loggedOut()
                dispatch(push('/'))
                resolve()
            }
        })
    })
}

export const loadNextSession = (userKey) => dispatch => {

    const fetchSuccess = compose(dispatch, exercisesFetchingSuccess, getVal)
    const fetchError = compose(dispatch, exercisesFetchingError)
    const doNothingForNoValueOrFetch = ifElse(
        compose(isNil, getVal),
        () => {
        },
        fetchSuccess
    )

    return loadNextSessionForUser(userKey)
        .then(doNothingForNoValueOrFetch)
        .catch(fetchError)
}
