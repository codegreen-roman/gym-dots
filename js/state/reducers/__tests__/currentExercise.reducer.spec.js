import { currentExercise } from '../currentExercise.reducer'
import { WORKOUT_STATUS, SET_SUCCESSFULLY_COMPLETED, SET_FAILED } from '../../actions/types'

const action = {
    type: WORKOUT_STATUS,
    payload: {
        name: 'Monday Sweat',
        status: 'started',
        restTime: 60,
        sets: 5,
        setsLeft: 5,
        reps: 12,
        weight: 100,
        results: []
    }
}

test('currentExercise returns an object with specific fields', () => {
    const output = currentExercise({}, action)
    expect(output).toEqual(
        expect.objectContaining({
            name: 'Monday Sweat',
            status: 'started',
            restTime: expect.any(Number),
            sets: expect.any(Number),
            setsLeft: expect.any(Number),
            reps: expect.any(Number),
            weight: expect.any(Number),
            results: expect.any(Array)
        })
    )
})

test('currentExercise output matches the snapshot', () => {
    expect(currentExercise({}, action)).toMatchSnapshot()
})

describe('Workout in progress', () => {

    const state = {
        'exerciseKey': '',
        'restTime': 90,
        'name': 'Push-ups',
        'setsLeft': 2,
        'reps': 10,
        'weight': 0,
        sets: 5,
        results: [true, true, true]
    }

    test('registering result : Done', () => {

        const action = {
            type: SET_SUCCESSFULLY_COMPLETED,
            payload: {}
        }
        expect(currentExercise(state, action)).toMatchSnapshot()
    })

    test('registering result : Failed', () => {
        const action = {
            type: SET_FAILED,
            payload: {}
        }
        expect(currentExercise(state, action)).toMatchSnapshot()
    })
})
