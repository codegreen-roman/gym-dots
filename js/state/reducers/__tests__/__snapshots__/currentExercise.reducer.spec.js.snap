// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout in progress registering result : Done 1`] = `
Object {
  "exerciseKey": "",
  "name": "Push-ups",
  "reps": 10,
  "restTime": 90,
  "results": Array [
    true,
    true,
    true,
    true,
  ],
  "sets": 5,
  "setsLeft": 1,
  "weight": 0,
}
`;

exports[`Workout in progress registering result : Failed 1`] = `
Object {
  "exerciseKey": "",
  "name": "Push-ups",
  "reps": 10,
  "restTime": 90,
  "results": Array [
    true,
    true,
    true,
    false,
  ],
  "sets": 5,
  "setsLeft": 1,
  "weight": 0,
}
`;

exports[`currentExercise output matches the snapshot 1`] = `
Object {
  "name": "Monday Sweat",
  "reps": 12,
  "restTime": 60,
  "results": Array [],
  "sets": 5,
  "setsLeft": 5,
  "status": "started",
  "weight": 100,
}
`;
